test/e2e/storage/csimock/csi_honor_pv_reclaim_policy.go

```diff
@@ -0,0 +1,192 @@
+			framework.ExpectNoError(err, "failed to wait for PV to be deleted")
+
+			ginkgo.By(fmt.Sprintf("Verifying that the driver received DeleteVolume call for PV %s", pv.Name))
+			gomega.Expect(m.driver.GetCalls(ctx)).To(gomega.ContainElement(gomega.HaveField("Method", gomega.Equal("DeleteVolume"))))
+		})
+
+		ginkgo.It("Dynamic provisioning should honor pv retain reclaim policy", func(ctx context.Context) {
+			m.init(ctx, testParameters{
+				registerDriver:             true,
+				enableHonorPVReclaimPolicy: true,
+				reclaimPolicy:              ptr.To(v1.PersistentVolumeReclaimRetain),
+			})
+			ginkgo.DeferCleanup(m.cleanup)
+
+			_, pvc := m.createPVC(ctx)
+
+			ginkgo.By(fmt.Sprintf("Waiting for PVC %s to be bound", pvc.Name))
+			pvs, err := e2epv.WaitForPVClaimBoundPhase(ctx, f.ClientSet, []*v1.PersistentVolumeClaim{pvc}, framework.ClaimProvisionTimeout)
+			framework.ExpectNoError(err, "failed to wait for PVC to be bound")
+			gomega.Expect(pvs).To(gomega.HaveLen(1), "expected 1 PV to be bound to PVC, got %d", len(pvs))
+
+			pv := pvs[0]
+			ginkgo.By(fmt.Sprintf("PVC %s is bound to PV %s", pvc.Name, pv.Name))
+			gomega.Expect(pv.Spec.PersistentVolumeReclaimPolicy).To(gomega.Equal(v1.PersistentVolumeReclaimRetain),
+				"expected PV %s to have reclaim policy %s, got %s", pv.Name, v1.PersistentVolumeReclaimRetain, pv.Spec.PersistentVolumeReclaimPolicy)
+
+			ginkgo.By(fmt.Sprintf("Verifying that the PV %s does not have finalizer %s after creation", pv.Name, storagehelpers.PVDeletionProtectionFinalizer))
+			gomega.Consistently(ctx, framework.GetObject(f.ClientSet.CoreV1().PersistentVolumes().Get, pv.Name, metav1.GetOptions{})).
+				WithPolling(framework.Poll).WithTimeout(framework.ClaimProvisionTimeout).ShouldNot(gomega.HaveField("Finalizers",
+				gomega.ContainElement(storagehelpers.PVDeletionProtectionFinalizer)), "failed to wait for PV to not have finalizer %s", storagehelpers.PVDeletionProtectionFinalizer)
+
+			ginkgo.By(fmt.Sprintf("Deleting PV %s", pv.Name))
+			err = f.ClientSet.CoreV1().PersistentVolumes().Delete(ctx, pv.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete PV %s", pv.Name)
+
+			ginkgo.By(fmt.Sprintf("Deleting PVC %s", pvc.Name))
+			err = f.ClientSet.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete PVC %s", pvc.Name)
+
+			ginkgo.By(fmt.Sprintf("Waiting for PV %s to be deleted", pv.Name))
+			err = e2epv.WaitForPersistentVolumeDeleted(ctx, f.ClientSet, pv.Name, framework.Poll, 2*time.Minute)
+			framework.ExpectNoError(err, "failed to wait for PV to be deleted")
+
+			ginkgo.By(fmt.Sprintf("Verifying that the driver did not receive DeleteVolume call for PV %s", pv.Name))
+			gomega.Expect(m.driver.GetCalls(ctx)).NotTo(gomega.ContainElement(gomega.HaveField("Method", gomega.Equal("DeleteVolume"))))
```

Comment:
I don't see why we need to bother about testing delete when reclaim policy is retain. It's expected by design not to call DeleteVolume even before my change.

Reply:
There's no an existing e2e test that can verify the deletion behavior of a pv with a retain reclaim policy. So I added a new e2e test for this case. Another reason is https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go#L1167. May we should add another new e2e test for this case that the pv changes its reclaim policy from delete to retain?

Reply:
let's make that in another PR, for now this is good enough.